### PR TITLE
Resend push log after logged in

### PR DESCRIPTION
**Pull request by dhei**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Implement resend push log after a new auth token is received.

### CODE REVIEW

File: AppCenterPush/AppCenterPush/MSPush.m [hunk 1]
```diff
@@ -389,4 +392,10 @@ - (BOOL)didReceiveRemoteNotification:(NSDictionary *)userInfo fromUserNotificati
   return NO;
 }
 
+- (void)authTokenContext:(__unused MSAuthTokenContext *)authTokenContext didReceiveAuthToken:(__unused NSString *)authToken {
+  if (self.pushToken) {
```
Comment (guperrot) [inline AppCenterPush/AppCenterPush/MSPush.m]: Unless you can guarantee it's always same thread accessing pushToken (and that would deserve comments), accessing self.pushToken after a if is not thread safe.
Either mark pushToken atomic and use a local variable before the if or use synchronized.

Comment (annakocheshkova) [inline AppCenterPush/AppCenterPush/MSPush.m]: This will execute if Push is disabled.

Comment (annakocheshkova) [reply AppCenterPush/AppCenterPush/MSPush.m]: We decided that the best place to add / remove delegate is in `applyEnabledState`

File: AppCenterPush/AppCenterPushTests/MSPushTests.m [hunk 2]
```diff
@@ -60,6 +62,9 @@ @implementation MSPushTests
 - (void)setUp {
   [super setUp];
   [MSUserIdContext resetSharedInstance];
+
+  // Need to reset defaults because enabled state gets persisted on disk otherwise, causing test inconsistencies.
```
Comment (dhei) [inline AppCenterPush/AppCenterPushTests/MSPushTests.m]: This comment is out of date, should be removed.

File: AppCenterPush/AppCenterPush/MSPush.m [hunk 3]
```diff
@@ -389,4 +392,13 @@ - (BOOL)didReceiveRemoteNotification:(NSDictionary *)userInfo fromUserNotificati
   return NO;
 }
 
+- (void)authTokenContext:(__unused MSAuthTokenContext *)authTokenContext didReceiveAuthToken:(__unused NSString *)authToken {
+
+  // Make a copy of push token so that this code is thread safe.
```
Comment (guperrot) [inline AppCenterPush/AppCenterPush/MSPush.m]: pointer copy, not value (we don't the copy flag on the property), I think the comment is confusing.
we could be more implicit like use a local reference to be thread safe?

Comment (dhei) [reply AppCenterPush/AppCenterPush/MSPush.m]: Fixed by adding `copy` attribute to `pushToken` property.

File: AppCenter/AppCenterTests/MSAuthTokenContextTests.m [hunk 4]
```diff
@@ -45,4 +45,18 @@ - (void)testSetAuthToken {
...
+  id<MSAuthTokenContextDelegate> delegateMock = OCMProtocolMock(@protocol(MSAuthTokenContextDelegate));
+  [self.sut addDelegate:delegateMock];
+
+  // Then
+  OCMReject([delegateMock authTokenContext:self.sut didReceiveAuthToken:OCMOCK_ANY]);
+
+  // When
+  [self.sut removeDelegate:delegateMock];
+  self.sut.authToken = @"something";
+}
```
Comment (ElektrojungeMS) [inline AppCenter/AppCenterTests/MSAuthTokenContextTests.m]: it's not obvious to me why we are setting the auth token after removing a delegate. I also don't see an actual test after `// When` or maybe I'm missing something?

File: AppCenterPush/AppCenterPush/Internal/Model/MSPushLog.h [hunk 5]
```diff
@@ -7,6 +7,6 @@
 /**
  * Push token for push service
  */
-@property(nonatomic) NSString *pushToken;
+@property(nonatomic, copy) NSString *pushToken;
```
Comment (ElektrojungeMS) [inline AppCenterPush/AppCenterPush/Internal/Model/MSPushLog.h]: Nice!

200w_d

File: AppCenter/AppCenterTests/MSAuthTokenContextTests.m (hunk 4 above)
Comment (ElektrojungeMS) [reply AppCenter/AppCenterTests/MSAuthTokenContextTests.m]: Gosh, I just realized how the test works! Sorry about my comment.

File: AppCenterPush/AppCenterPush/MSPush.m (hunk 3 above)
Comment (guperrot) [reply AppCenterPush/AppCenterPush/MSPush.m]: Ok I was thinking of just fixing the comment instead actually, not sure you need a value copy here.

File: AppCenterPush/AppCenterPush/Internal/MSPushPrivate.h [hunk 6]
```diff
@@ -17,14 +18,14 @@ static NSString *const kMSPushNotificationOldCustomDataKey = @"mobile_center";
...
+@interface MSPush () <NSUserNotificationCenterDelegate, MSAuthTokenContextDelegate>
 #else
-@interface MSPush ()
+@interface MSPush () <MSAuthTokenContextDelegate>
 #endif
 
 @property(nonatomic) id<MSPushDelegate> delegate;
 
-@property(nonatomic) NSString *pushToken;
+@property(atomic) NSString *pushToken;
```
Comment (guperrot) [inline AppCenterPush/AppCenterPush/Internal/MSPushPrivate.h]: Then according to https://github.com/Microsoft/AppCenter-SDK-Apple/pull/1309/files#r258166953 this one should be copy too.

File: AppCenterPush/AppCenterPush/MSPush.m (hunk 3 above)
Comment (guperrot) [reply AppCenterPush/AppCenterPush/MSPush.m]: Synced offline, should be copy for best practice.

File: AppCenterPush/AppCenterPush/Internal/MSPushPrivate.h (hunk 6 above)
Comment (dhei) [reply AppCenterPush/AppCenterPush/Internal/MSPushPrivate.h]: Good catch, thanks a lot.

File: AppCenterPush/AppCenterPushTests/MSPushTests.m [hunk 7]
```diff
@@ -60,6 +62,8 @@ @implementation MSPushTests
 - (void)setUp {
   [super setUp];
   [MSUserIdContext resetSharedInstance];
+
```
Comment (jwallra) [inline AppCenterPush/AppCenterPushTests/MSPushTests.m]: Remove extra new line

File: AppCenterPush/AppCenterPush/Internal/MSPushPrivate.h (hunk 6 above)
Comment (ElektrojungeMS) [reply AppCenterPush/AppCenterPush/Internal/MSPushPrivate.h]: yes, it should! Sorry for missing this!

**Pull request merged**